### DISCUSSION TITLE
Stage Progression and Terraform "undeclared variable" warnings

### DISCUSSION THREAD

**Discussion by ranger-zero** [GoogleCloudPlatform/cloud-foundation-fabric#2028] [General]
![Screenshot 2024-01-30 7 26 06 AM](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/assets/136013030/ebbf60ce-a030-459d-ba6e-51fa575ec83d)
I would like to understand if terraform warnings on "undeclared variables" based on previous stage json output should be of concern and if so defined in my current stages variables.tf.

Case: I completed the bootstrap stage and moved to resman stage. I imported the "0-bootstrap.auto.tfvars.json" via gcloud alpha storage command. I updated my tfvars with my custom configuration and ran terraform init, plan, then apply. In my terraform apply I get the "Warning: Value for undeclared variable" due to the "logging" variable defined in my json but no variable block defined. 

I am able to move past the warning but I am wondering if this should be defined or not. I know the "logging" variables created in bootstrap stage and passed via the json are the default created in bootstrap stage. 

Thoughts from the community?

**Comment by ludoo** [comment]
This is completely fine, if a bit disconcerting at first. :)

The FAST stage model leverages the concept of interfaces, where producers -- earlier stages -- spit out data (an output interface) which is guaranteed to always exist, and consumers -- later stages in the chain -- consume whatever they need from it (their input interface).

There's very little value and a lot of overhead in forcing consumers to accept the entire output interface, so we don't do that and this is the reason for the warnings your are seeing. Just ignore them. :)

**Reply by ranger-zero** [reply]
This pattern makes sense and I appreciate the quick response. And great work to the team of contributors on this. I am building out my startups infrastructure while understanding these concepts and applying these in an enterprise setting. I am very interested in the core services of the data platform and gke. Thank you!

**Reply by ludoo** [reply]
Thank you! Anydoubts you have don't hesitate to open a discussion or issue. :) Good luck with your startup!

**Comment by markjons2** [comment]
"Forcing consumers to accept the whole [output interface](https://socialhubin.com/why-i-cant-see-comments-on-threads-problem-solved/) isn't very useful and can cause a lot of unnecessary work, so we avoid doing that. That's why you're seeing those warnings. You can safely ignore them."
